File: cli/cmd/pastes.ts
import { Command } from "https://deno.land/x/cliffy@v1.0.0-rc.2/command/command.ts";
import supabaseClient from "../util/supabase.ts";
import {
  Session,
  User,
} from "https://esm.sh/v130/@supabase/gotrue-js@2.46.1/dist/module/index.js";
import { getPasteUrl } from "../util/helpers.ts";

const PastesCommand = new Command()
  .name("pastes")
  .description("List your pastes")
  .alias("ls")
  .alias("list")
  .action(async () => {
    const sessionData = localStorage.getItem("session");
    const userData = localStorage.getItem("user");

    if (!sessionData || !userData) {
      console.error("You are not logged in.");
      Deno.exit(1);
    }

    const user: User = JSON.parse(userData);
    const session: Session = JSON.parse(sessionData);

    const { data, error: sessionError } = await supabaseClient.auth.setSession({
      access_token: session.access_token,
      refresh_token: session.refresh_token,
    });

    if (sessionError) {
      console.error(sessionError);
      Deno.exit(1);
    }

    const { data: pastes, error } = await supabaseClient
      .from("pastes")
      .select("*")
      .eq("author", data.user?.id);

    if (error) {
      console.error(error);
      Deno.exit(1);
    }

    if (!pastes) {
      console.log("You have no pastes.");
      Deno.exit(0);
    }

    console.log("Your pastes:");

    for (const paste of pastes) {
      console.log(
        `- ${paste.title ?? "Untitled Paste"}: ${getPasteUrl(paste.id)} [${
          paste.private ? "Private" : "Public"
        }]`,
      );
    }
  });

export default PastesCommand;

File: cli/cmd/pastes.go
package cmd

import (
	"fmt"
	"time"

	"github.com/urfave/cli/v2"
)

type CustomTime struct {
	time.Time
}

const customLayout = "2006-01-02T15:04:05.999999"

// Implement the UnmarshalJSON method for CustomTime
func (ct *CustomTime) UnmarshalJSON(data []byte) error {
	// Remove the surrounding quotes from the JSON string
	if string(data) == "null" {
		return nil
	}
	trimmedData := data[1 : len(data)-1]
	parsedTime, err := time.Parse(customLayout, string(trimmedData))
	if err != nil {
		return err
	}
	ct.Time = parsedTime
	return nil
}

type Paste struct {
	ID          string     `json:"id"`
	CreatedAt   CustomTime `json:"created_at,omitempty"`
	Author      string     `json:"author"`
	File        string     `json:"file"`
	Draft       bool       `json:"draft"`
	ExpiresAt   CustomTime `json:"expires_at,omitempty"`
	Title       string     `json:"title"`
	Description string     `json:"description"`
	Language    string     `json:"language"`
}

var PastesCommand = cli.Command{
	Name:    "pastes",
	Aliases: []string{"ls", "list"},
	Usage:   "Manage your pastes.",
	Action: func(cCtx *cli.Context) error {
		user, err := supabase.Auth.User(ctx, settings.AccessToken)

		if err != nil || user == nil {
			cli.Exit("You don't seem to be signed in. Try running `openbin login` to sign in.", 1)
		}

		supabase.DB.AddHeader("Authorization", fmt.Sprintf("Bearer %s", settings.AccessToken))

		var pastes []Paste
		err = supabase.DB.From("pastes").Select("*").Eq("author", user.ID).Execute(&pastes)

		if err != nil {
			cli.Exit("Could not get the pastes.", 1)
			return err
		}

		for _, paste := range pastes {
			visibility := "Public"

			if paste.Draft {
				visibility = "Draft"
			}

			title := paste.Title

			if title == "" {
				title = "Untitled Paste"
			}

			fmt.Printf("\n- %s: %s [%s]\n", title, fmt.Sprintf("https://openbin.dev/paste/%s", paste.ID), visibility)
		}

		return nil
	},
}
